### PR TITLE
DB-9048 Support empty and cast default values

### PR DESCRIPTION
**Pull request by arnaud-splice**

### CODE REVIEW

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/EmptyDefaultConstantNode.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+import com.splicemachine.db.iapi.services.compiler.MethodBuilder;
+import com.splicemachine.db.iapi.services.sanity.SanityManager;
+import com.splicemachine.db.iapi.error.StandardException;
+import com.splicemachine.db.iapi.types.DataValueDescriptor;
+import com.splicemachine.db.iapi.types.DataTypeDescriptor;
+
+import java.util.List;
+import java.util.Vector;
+/**
+ * An UntypedNullConstantNode represents a SQL NULL before it has
```
Comment (yxia92) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/EmptyDefaultConstantNode.java]: Please correct the comments that refer to UntypedNullConstantNode.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/EmptyDefaultConstantNode.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+ * been bound.  The bind() operation will replace the UntypedNullConstantNodes
+ * with typed ConstantNodes.
+ */
+
+public final class EmptyDefaultConstantNode extends ConstantNode
+{
+    /**
+     * Constructor for an UntypedNullConstantNode.  Untyped constants
```
Comment (yxia92) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/EmptyDefaultConstantNode.java]: Correct comment referring to UntypedNullConstantNode.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/EmptyDefaultConstantNode.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+     * we shouldn't make it to generate
+     *
+     * @param acb	The ExpressionClassBuilder for the class being built
+     * @param mb	The method the expression will go into
+     */
+    void generateConstant(ExpressionClassBuilder acb, MethodBuilder mb)
+    {
+        if (SanityManager.DEBUG)
+        {
+            SanityManager.THROWASSERT("generateConstant() not expected to be called for UntypedNullConstantNode because we have implemented our own generateExpression().");
```
Comment (yxia92) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/EmptyDefaultConstantNode.java]: Same here, correct the error messages.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/EmptyDefaultConstantNode.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+     * Translate a Default node into a default value, given a type descriptor.
+     *
+     * @param typeDescriptor	A description of the required data type.
+     *
+     */
+    public DataValueDescriptor convertDefaultNode(DataTypeDescriptor typeDescriptor)
+            throws StandardException
+    {
+        /*
+         ** The default value is null, so set nullability to TRUE
```
Comment (yxia92) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/EmptyDefaultConstantNode.java]: This comment is for UntypedNullConstantNode, please remove.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/EmptyDefaultConstantNode.java (hunk 1 above)
Comment (arnaud-splice) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/EmptyDefaultConstantNode.java]: Done

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/EmptyDefaultConstantNode.java (hunk 2 above)
Comment (arnaud-splice) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/EmptyDefaultConstantNode.java]: Done

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/EmptyDefaultConstantNode.java (hunk 3 above)
Comment (arnaud-splice) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/EmptyDefaultConstantNode.java]: Done

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/EmptyDefaultConstantNode.java (hunk 4 above)
Comment (arnaud-splice) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/EmptyDefaultConstantNode.java]: Done

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Pull request merged**